Bring watchdog feature branch into master

fkjagodzinski opened the pull request:
### Description

<!--
    Required
    Add here detailed changes summary, testing results, dependencies
    Good example: https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html (Pull request template)
-->
Bring `feature-watchdog` into `master`.

### Pull request type

<!--
    Required
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front
    of them would change the meaning incorrectly. The only changes to be made are to add a description text under the
    description heading and to add a 'x' to the correct box.
-->
    [ ] Fix
    [ ] Refactor
    [ ] Target update
    [x] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

### Reviewers

<!--
    Optional
    Request additional reviewers with @username
-->
@rajkan01 @donatieng @jamesbeyond @0xc0170

### Release Notes

<!--
    Optional
    In case of breaking changes, functionality changes or refactors, please add release notes here. 
    For more information, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html#pull-request-types).
-->

Adding new HAL API: `watchdog` and `reset_reason`
Adding new drivers: `ResetReason` and `Watchdog`

`ResetReason` -  checking the cause of the reset
`Watchdog` - watchdog timer to reset the system

Commit:
Add Watchdog HAL API specification headers

HAL watchdog functionality will be implemented as two separate APIs. The reset
reason API allows a user to detect the last system reset reason to identify if
a Watchdog was triggered. The Watchdog API allows configuring and updating
Watchdog timers on all boards. This commit defines the headers.

Commit:
Add K64F reset reason reference implementation

Commit:
Add simple watchdog reference implementation for K64F

- Add preprocessor guard to watchdog api that errors if the reset reason api is not also implemented
- Add RESET_REASON and WATCHDOG to K64F targets.json
- Add watchdog reference implementation

Commit:
Fix watchdog API issues

- Fix typo in module comment
- Redefine the default system behaviour in sleep mode
- Guard K64F enableWait flag
- Remove bit shifts from reset reason enum

fkjagodzinski committed:
Tests: HAL API: Watchdog: Add tests

Commit:
Add Reset Reason platform API

Commit:
Add reset reason reference implementation STM32

Commit:
Add function to fetch platform specific reset reason register values

Commit:
Add independent watchdog reference implementation for STM32

Commit:
Amend reset reason driver API
- Change API to match C++ API throughout drivers
- Amend HAL API documentation to be more specific

Commit:
Add Watchdog driver API

fkjagodzinski committed:
Tests: HAL API: Reset_reason: Add tests

Commit:
Remove window and sleep mode options for watchdog API

fkjagodzinski committed:
Tests: HAL API: Watchdog: Add dev reset tests

Commit:
Add missing license headers

Commit:
Move watchdog parameter validation into the driver layer

Commit:
Fix Watchdog::stop on K64F target

WDOG_Disable will suspend the watchdog until a reset is triggered instead of stopping it altogether. Deinit will disable it until it is reinitialized.

fkjagodzinski committed:
K64F: watchdog HAL: Fix max_timeout.

Limit MAX_TIMEOUT_MS to UINT32_MAX.
Improve MAX_TIMEOUT_MS handling.
Fix hal_watchdog_get_platform_features() returning
.max_timeout = 0xfffffff8 instead of 0xffffffff.

fkjagodzinski committed:
Tests: HAL API: Watchdog: Add time accuracy tests

Commit:
Add RESET_REASON and WATCHDOG definitions to hexiwear build target

Commit:
Rename watchdog.c -> watchdog_api.c to prevent name collision with Platform API

fkjagodzinski committed:
Tests: Drivers: Reset_reason: Add tests

fkjagodzinski committed:
Tests: Drivers: Watchdog: Add tests

bulislaw committed:
Add definition of reset reason and watchdog doxy groups

fkjagodzinski committed:
STM32F4: watchdog HAL: Fix issues found with tests

Fix WATCHDOG_STATUS_INVALID_ARGUMENT for timeout values from 1 ms to
407 ms (fix inability to set prescaler bits to zero).
Fix timeout <-> IWDG registers conversions.
Fix read & write access to IWDG_PR and IWDG_RLR registers.
Fix LSI RC frequency setting.
Limit MAX_TIMEOUT_MS to UINT32_MAX.

fkjagodzinski committed:
K64F: watchdog HAL: Fix init() and stop()

Added a missing wait for the WCT window end. Without it, consecutive
init() or stop() calls were ignored.

fkjagodzinski committed:
STM: HAL: Reset_reason: Correct return values

bulislaw committed:
Add RESET_REASON and WATCHDOG to doxygen options

bulislaw committed:
Document reset reason enum

fkjagodzinski committed:
Tests: HAL: Watchdog: Fix timing accuracy test

Relax time measurement margins for UBLOX_EVK_ODIN_W2.

bulislaw committed:
Disable Watchdog on Odin as it fails intermittently

jeromecoutant committed:
STM32 Watchdog : move API file to STM family level

Commit:
Add Watchdog and ResetReason headers to mbed.h

Adding headers to mbed.h will prevent you having to manually include
them when using the drivers in an application.

stevew817 committed:
Add support for watchdog on Silicon Labs devices

Commit:
Add text to watchdog documentation indicating that it will continue to run in sleep modes

jeromecoutant committed:
STM32 WATCHDOG : compilation issue with typed define

stevew817 committed:
Add implementation of reset reason on Silicon Labs parts

jeromecoutant committed:
STM32 WATCHDOG : update STM32L0 HAL_IWDG_Init to a newest version

jeromecoutant committed:
STM32 WATCHDOG : use ST HAL in order to make code commun for all STM32

Commit:
NCS36510: Added watchdog implementation.

jeromecoutant committed:
STM32 WATCHDOG : increase timeout value

fkjagodzinski committed:
Tests: Watchdog: Update test case teardown

Added a mechanism which will prevent an initialised watchdog from
resetting the device during final greentea communication.
This allows testing watchdog timeouts as short as 50 ms.

fkjagodzinski committed:
Tests: Watchdog: Decrease default timeout to 100 ms

fkjagodzinski committed:
Tests: Watchdog: Update timing test

Skip test cases with unsupported timeout values.

Commit:
fix bugs introduced on rebase

ccli8 committed:
[Nuvoton] Support reset reason

ccli8 committed:
[Nuvoton] Support watchdog timer

ccli8 committed:
[NUC472] Fix WDT driver in BSP

ccli8 committed:
Enlarge wait time for flushing serial buffer in Greentea test code

Original wait time is not enough for UART to flush out before deepsleep/reset on some targets.

ccli8 committed:
Fix NUMAKER_PFM_NANO130 cannot pass mbed_hal/watchdog_reset test

On NUMAKER_PFM_NANO130 target, WDT's clock source is fixed to LIRC, which is much
less accurate than other targets. Enlarge delta define to pass this test.

Commit:
Added reset_reason feature for TMPM066 & TMPM46B

Commit:
Added reset_reason feature for TMPM3H6 & TMPM4G9

Commit:
test(api-watchdog): Fix errors in compiling watchdog greentea tests

The latest rebase of the watchdog feature branch introduced errors in compiling
watchdog tests due to missing headers.

- Watchdog HAL API test
  Include missing header files to main.cpp (mbed_wait_api.h, stdlib.h).

- Watchdog HAL API timing test
  Include missing header files to main.cpp (us_ticker_api.h).

- Watchdog Driver API test
  Include missing header files to main.cpp (mbed_wait_api.h, stdlib.h).
  Inject mbed namespace into main.cpp to fix Watchdog name resolution.

Commit:
fix(hal-watchdog): Add feature guards to K64F watchdog implementation

Commit:
Update license headers

Commit:
Apply new astyle config style

Commit:
Fix more astyle failures

fkjagodzinski committed:
Tests: Watchdog: Fix config access in the host script

Move the calls to get_config_item() to the setup() method. The config is
not available in the initializer yet.

fkjagodzinski committed:
Tests: Watchdog: Update the reset-to-sync delay

Change the config parameter used as a delay before sending the sync
packet after the device reset in watchdog and reset_reason tests. Use
'forced_reset_timeout' instead of 'program_cycle_s'.

Commit:
Added reset_reason feature for TMPM3HQ

Commit:
Edit ResetReason_tests.h

Make minor grammar edits.

fkjagodzinski committed:
Docs: Watchdog: Update code comments

* capitalize driver class name,
* reword test docs for the driver & HAL,
* capitalize Mbed name,
* reword the comments explaining the SERIAL_FLUSH_TIME_MS macro.

Commit:
small grammatical changes

Not sure if the formatting's still alright on these.

Commit:
Capitalizing Watchdog

Commit:
capitalizing class name

Hope this is alright.

Commit:
fix: Remove excess comma from device_has key in targets.json

Commit:
fix(hal-watchdog): Remove 'RESET_REASON' flags from unsupported devices

Commit:
fix(hal-watchdog): Add 'RESET_REASON' flags to supported devices

Commit:
fix(hal-watchdog): Guard TMPM 'ResetReason' code behind feature flag

rajkan01 committed:
Watchdog refactoring to multithreaded thread.

-Added API to register muliple threads to watchdog drivers
-Watchdog timeout reconfigures everytime whenever new register thread with longer timeout period
-New APIs for watchdog
 wd_register(const osThreadId_t tid, const uint32_t timeout) to register to watchdog
 wd_unregister(const osThreadId_t tid) to unregister to watchdog
 kick(const osThreadId_t tid) to refresh the watchdog

rajkan01 committed:
Add HwWatchdog

    - mbed_watchdog_mgr has interface name mbed_wdog_manager_start(),mbed_wdog_manager_stop(),mbed_wdog_manager_kick()
    - HwWatchdog is going to attach with LowPowerTIcker for periodic callback functionality
    - mbed_wdog_manager_start() will either get start either by BL/RTOS Aps,it reads the timeout value specified via macro and macro gets defined in target.json file.
    - mbed_wdog_manager_start() internally configure below HAL hw watchdog with timeout specified via target.json
    - mbed_wdog_manager_start() internally divide the timeout(specified in target.json) by the 2 and attach LowPowerTicker with periodic callback of hw_kick()
    - mbed_wdog_manager_start() internally create one instance of sw watchdog class,to access the static list data structure of sw watchdog class
    - mbed_wdog_manager_kick() function periodically get called and refresh the hw watchdog to avoid watchdog reset
    - converted C++ code into C based APIs
    - added boolean to control watchdog start and stop
    - Added detach from ticker on stop API

rajkan01 committed:
Add SW Watchdog

        -SW watchdog has interface name start(),stop(),kick()	Sw watchdog internally has static list and shared across multiple instance of SW watchdog
	- Sw watchdog initialize timeout value,unique string via constructor whenever threads created sw watchdog object
	-Threads make sure pass proper timeout value,Unique string while creating the instance.
	-start() called by components(BLE,WIFI etc.,),it adds the entry into static list with few details current count ,etc.,
	-kick() called by registered components(BLE,WIFI etc.) to reset current count to zero.
        -is_alive - interface API to mbed_watchdog_manager
        -implementation optimization

rajkan01 committed:
HW Watchdog and SW watchdog interface

  - Added Hw watchdog periodic kick calls software watchdog is_alive check

rajkan01 committed:
Hw Watchdog UnitTest cases

    -Added the unit test case for testing Hw watchdog
    -Added the supported stubs files

rajkan01 committed:
SW watchdog unittest cases

  -Added the mock class function to mock mbed_assert_internal
  -Added the unit test case to test start,kick,stop
  -Modified the interface api name from is_alive to process
  -added the unit test cases for process

fkjagodzinski committed:
Tests: Watchdog: Remove outdated driver tests

The Watchdog driver has been updated making these tests incompatible
with the new API. The Watchdog driver functionality is tested with the
unit tests located in UNITTESTS/drivers/Watchdog/ .

fkjagodzinski committed:
Tests: Watchdog: Add Watchdog Manager tests

rajkan01 committed:
fix: Fixing rebased created bugs

rajkan01 committed:
update:Remove DEVICE_WATCHDOG and added RESET_RESON in stm32 family

rajkan01 committed:
doxy spell check fix for watchdog

rajkan01 committed:
fix for build failure issues

rajkan01 committed:
Revert "fix for build failure issues"

This reverts commit 3ce1d4940cb7fd9d74ca2be75ed08b76854ff534.

rajkan01 committed:
Removed RESET_REASON for NRF52832,NRF52840 family and Removed WATCHDOG for H743ZI board

rajkan01 committed:
fix for Jenkin build issue

rajkan01 committed:
Trailing space issue fix

rajkan01 committed:
Added the SPDX identifier

rajkan01 committed:
Removed RESET_REASON for NRF51_DK board

rajkan01 committed:
Fix for the Ci build issue

rajkan01 committed:
fix for spell check and header file

rajkan01 committed:
fix for astyle

rajkan01 committed:
Incorporated the Reivew Comments

rajkan01 committed:
Incorporated the Reivew Comments

rajkan01 committed:
Update Watchdog

 - Changed the process into static method
 - used the singletonptr for creating the low power ticker instance
 - Added the mbed stub into cmake build for cellularnonipsocket,loramacrypto

rajkan01 committed:
Added the ticker support
 -Modified the device dont have support of lp ticker will ticker for watchdog callback register

rajkan01 committed:
Fixed the unittest case failure

rajkan01 committed:
Fix for Problem copying the instance from BSS to stack
  -Created global instance of either LowPowerTicker or Ticker

fkjagodzinski committed:
Tests: ResetReason: Use the HAL watchdog API

The Watchdog driver has been updated making this test incompatible
with the new API. Use HAL API instead.

fkjagodzinski committed:
Test: update HAL reset_reason test

Update the watchdog timing requirements:
1. The watchdog should trigger at, or after the timeout value.
2. The watchdog should trigger before twice the timeout value.

fkjagodzinski committed:
Test: update HAL watchdog test

Update the watchdog timing requirements:
1. The watchdog should trigger at, or after the timeout value.
2. The watchdog should trigger before twice the timeout value.

fkjagodzinski committed:
Test: update HAL watchdog-reset test

Update the watchdog timing requirements:
1. The watchdog should trigger at, or after the timeout value.
2. The watchdog should trigger before twice the timeout value.
Remove explicit calls to sleep.

fkjagodzinski committed:
Test: update HAL watchdog-timing test

Update the watchdog timing requirements:
1. The watchdog should trigger at, or after the timeout value.
2. The watchdog should trigger before twice the timeout value.

fkjagodzinski committed:
Remove RESET_REASON from MCU_NRF52840

fkjagodzinski committed:
Tests: Update Watchdog Manager test

Update the watchdog timing requirements:
1. The watchdog should trigger at, or after the timeout value.
2. The watchdog should trigger before twice the timeout value.

fkjagodzinski committed:
Test: update reset_reason driver test

Update the watchdog timing requirements:
1. The watchdog should trigger at, or after the timeout value.
2. The watchdog should trigger before twice the timeout value.

Use a SERIAL_FLUSH_TIME_MS value consistent with other tests.

0xc0170 committed:
Merge pull request #10645 from fkjagodzinski/rajkan01-new_watchdog_api

Watchdog API update

fkjagodzinski committed:
Test: watchdog_mgr: Remove dead code

adbridge committed:
Merge pull request #10777 from fkjagodzinski/fix-watchdog-missing_commits

Bring back the missing watchdog commits

fkjagodzinski committed:
HAL: Watchdog: Update the header file

Add a SPDX license identifier.
Add defined & undefined behavior.

jamesbeyond reviewed (approved):
The file changes should have already got reviewed during the FB PR.
I had a quick scan through the files, I am happy about the changes.

bulislaw commented on drivers/Watchdog.h:
```diff
@@ -0,0 +1,120 @@
+
+namespace mbed {
+
+/** \addtogroup drivers */
```
@AnotherButler Hi Amanda, could you review the docs in the headers? We should improve the wording and grammar before merging it to master.

bulislaw commented on drivers/Watchdog.h:
```diff
@@ -0,0 +1,120 @@
+     * Calling this function will attempt to disable any currently running
+     * watchdog timers if supported by the current platform.
+     *
+     * Assert with out called start
```
What happens when HW doesn't support stopping the watchdog??

bulislaw commented on platform/mbed_watchdog_mgr.h:
```diff
@@ -0,0 +1,97 @@
+/* mbed Microcontroller Library
```
Why do we actually need the watchdog manager? It needs to be documented that it's not user facing API and shouldn't be used, pointing to Watchdog instead.

bulislaw commented on hal/reset_reason_api.h:
```diff
@@ -0,0 +1,110 @@
+    RESET_REASON_UNKNOWN         /**< Unknown or unreadable reset reason **/
+} reset_reason_t;
+
+/** Fetch the reset reason for the last system reset
```
Could we copy the defined/undefined behaviours from the handbook here, to much other hal apis.

0xc0170 replied:
Added in f084bd5a9dd85e7267dbc29e5d4a3530a6fbab9a

bulislaw commented on hal/watchdog_api.h:
```diff
@@ -0,0 +1,149 @@
+ * @{
+ */
+
+/** \file watchdog_api.h
```
Could we copy the defined/undefined APIs from the handbook to match other hal apis.

fkjagodzinski replied:
Added in https://github.com/ARMmbed/mbed-os/pull/10857/commits/a33536e78e9efe266eaa44ce6e4ec8b416b64bcd

bulislaw reviewed (changes requested):
Lots of comments in the code. Big points:
* Docs
* What's the watchdog manager?

fkjagodzinski committed:
HAL: ResetReason: Update the header file

Add defined & undefined behavior.
Minor doxygen fixes.

fkjagodzinski committed:
ResetReason: Update the header file

Minor doxygen fixes.

0xc0170 commented on TESTS/host_tests/sync_on_reset.py:
```diff
@@ -0,0 +1,73 @@
+"""
+mbed SDK
+Copyright (c) 2017 ARM Limited
+
```
SPDX identifier (might be in another PR that is in CI now)?

0xc0170 replied:
I'll fix this one

0xc0170 commented on hal/watchdog_api.h:
```diff
@@ -0,0 +1,149 @@
+/** \addtogroup hal */
+/** @{*/
+/* mbed Microcontroller Library
+ * Copyright (c) 2017 ARM Limited
```
as SPDX, also year should be 2018-2019 (if you do changes to files, pls update)

0xc0170 replied:
I'll fix this one

0xc0170 commented on platform/mbed_watchdog_mgr.cpp:
```diff
@@ -0,0 +1,110 @@
+ */
+static void mbed_wdog_manager_kick()
+{
+    core_util_critical_section_enter();
```
as requested below, is there documentation for this manager? looking at the implementation it kind of mixing hal with driver.

My question is : what is this bringing to the table? why is it not in driver?

0xc0170 commented on targets/targets.json:
```diff
@@ -47,6 +47,11 @@
                 "value": false
+            },
+            "hw-watchdog_timeout": {
+                "help": "Define the timeout in ms value LowPowerTicker to do HW kick",
```
is this low power ticker? I can see using wait with this macro and could not find reference using LowPowerTicker in this PR?

why does it have `macro_name` ? if it follows the default name , would be with prefix `MBED_CONF_TAR` or simarl to distinguish this comes from target

fkjagodzinski committed:
HAL: Watchdog: Update the header file

Minor doxygen fixes.

0xc0170 committed:
watchdog: fix copyright

0xc0170 committed:
watchdog: fix astyle issue

Missing a space after group and before license

0xc0170 committed:
watchdog hal: remove space

Commit:
Edit ResetReason.h

Edit file for active voice, complete sentences and grammar.

Commit:
Edit Watchdog.h

Edit file, mostly for active voice, consistent abbreviations, grammar and spelling.

Commit:
Edit watchdog_api.h

Edit file, mostly for active voice, consistent tense and international spelling.

Commit:
Edit reset_reason_api.h

Edit file, mostly for correct grammar.

0xc0170 committed:
Watchdog: refactor driver

Watchdog is hardware driver. It interacts with HAL - provides wrapper to interact with the peripheral.

Provides basic functionality: start/stop, get timeout/max timeout.
It is automatically kicked by a timer, according to the timeout set in ctor.

0xc0170 committed:
VirtualWatchdog: software watchdog

Refactor old Watchdog (it was not a driver) to become VirtualWatchdog.
This is software virtual watchdog. This it the primary used watchdog in user application.

VirtualWatchdog: has-a watchdog. Initializes hw watchdog - start it when first used, stops it when there is no more VirtualWatchdog in the system -
list is empty.

Adding also check to watchdog to make sure there is only one in the system - runtime error if multiple objects created to already
running hw watchdog.

0xc0170 committed:
Remove watchdog manager

Use VirtualWatchdog instead

0xc0170 committed:
mbed: add Watchdog and VirtualWatchdog

0xc0170 committed:
Watchdog: use get_instance to access watchdog hw

0xc0170 committed:
VirtualWatchdog: fix protection macro

0xc0170 committed:
Watchdog: cleanup after singleton usage

0xc0170 committed:
Watchdog: kick method should not have any external dep

0xc0170 committed:
Watchdog: fix dtor

0xc0170 committed:
watchdog: add WatchdogManager

Watchdog should be usable on it's own - kicking it via ticker. No Virtual involved.
VirtualWatchdog as well - services should use this one to have multiple one in the system.

There's WatchdogManager - basically internal class to manage ticker for Watchdog.

0xc0170 committed:
watchdog: remove manager

Watchdog can handle callbacks - VirtualManager can attach to the tick. This should simplify the logic.

Watchdog can tick on its own using tickers.
VirtualWatchdog uses attach to get a callback when Watchdog ticks - to process own
linked list of virtual watchdogs.

0xc0170 committed:
watchdog: fix astyle errors

fkjagodzinski committed:
Watchdog & VirtualWatchdog process() update

Add uint32 argument - pass how much time already passed to the callback.

0xc0170 committed:
Watchdog: clean-up start

0xc0170 committed:
VirtualWatchdog: make sure we are the only owner of watchdog

If virtual one is not the only owner we error - init error code in driver.

0xc0170 committed:
Watchdog: fix const static

Fixing error "non-static data member initializers only available with -std=c++11 or -std=gnu++11"

0xc0170 committed:
VirtualWatchdog: fix error code and timeout

0xc0170 committed:
error: add watchdog driver error

0xc0170 committed:
Watchdog: remove dead reference file

0xc0170 committed:
Watchdog: add missing docs doxygen

0xc0170 committed:
Watchdog: refactor timeout handling

Enable kicking directly by kick() and handle timeout by using private handler.

0xc0170 committed:
VirtualWatchdog: fix doxygen docs errors

0xc0170 committed:
Watchdog: fix callback - uses ms

0xc0170 committed:
Watchdog: fix timeout - use full range for watchdog

0xc0170 committed:
VirtualWatchdog: add config option to ctor to clarify the hw timeout

0xc0170 committed:
VirtualWatchdog: fix is hw init definition

0xc0170 committed:
VirtualWatchdog: fix name for member timeout

0xc0170 committed:
VirtualWatchdog: fix default timeout value (1ms is unreal)

0xc0170 committed:
Watchdog: fix include paths - use module name

0xc0170 committed:
Watchdog: keep it simple - providing simple driver functionality

No callbacks or tickers. User is handling kick() method directly based on timeout set.
VirtualWatchdog manages ticker and callback.

0xc0170 committed:
Watchdog: update unittests

0xc0170 committed:
unittest: rename watchdog to VirtualWatchdog

0xc0170 committed:
unittest: move watchdog files to drivers

Test VirtualWatchdog and Watchdog

0xc0170 committed:
unittest: remove platform watchdog (moved to drivers)

0xc0170 committed:
watchdog unittest: fix naming after refactor

0xc0170 committed:
unittest: add virtualwatchdog stub

0xc0170 committed:
VirtualWatchdog: fix missing kick and init check in process

Also syntax error for ticker definitions.

0xc0170 committed:
virtualWatchdog: fix alignment for tickers and one space

0xc0170 committed:
virtualWatchdog: fix alignment for tickers

0xc0170 committed:
Watchdog: example code - add kick()

fkjagodzinski committed:
Fix unittest errors

fkjagodzinski committed:
Fix more unittest errors

0xc0170 committed:
Watchdog: fix space in code example

fkjagodzinski committed:
Test: Watchdog: Add new driver tests

fkjagodzinski committed:
Watchdog: Fix timeout constraints

Update the start method to check 0 < timeout <= max_timeout.

fkjagodzinski committed:
Test: WatchdogManager: Remove tests

This component no longer exists in Mbed OS.

fkjagodzinski committed:
Test: Watchdog: Fix astyle errors

0xc0170 committed:
VirtualWatchdog: process fail test expects 0 as reset - no ticker running

0xc0170 committed:
Watchdog: add docs update about intention one vs multiple Watchdogs

VirtualWatchdog - multiple timers in the system - most used watchdog
Watchdog - simple driver for one system timer in the app - bootloader or similar

fkjagodzinski committed:
Test: Watchdog: Fix test_restart case

Watchdog::stop() is called in this test. If the target does not support
stopping the watchdog this case needs to be skipped.
Also improve handling ignored test cases.

0xc0170 committed:
VirtualWatchdog: fix doxygen ingroup and ctor

0xc0170 committed:
VirtualWatchdog: add comment about exclusive access to Watchdog

0xc0170 committed:
VirtualWatchdog: provide hw timeout via ctor

0xc0170 committed:
Watchdog: remove config value for timeout, set it dynamically

Use start method to set timeout

0xc0170 committed:
VirtualWatchdog: use timeout in ctor to set hw timeout

No config, neither additional parameter - the first instance of the class initalizes the peripheral.

fkjagodzinski committed:
Watchdog: Fix unittests

Compare get_timeout against the last timeout successfully set.
Merge most of the test cases into one. Do not rely on the tests
execution order.
Fix the watchdog_api_stub to keep timeout value set by init function.

0xc0170 committed:
Watchdog: remove VirtualWatchdog

We will provide documentation how to create your own VirtualWatchdog. It's simple,
create timeout and watchdog objects.

This class brought lot of discussion and questions. After our refactor we made this class
just a linked list of objects - something tickers can do as well (they already have linked list) and handling hw should be
done via Watchdog due to the limitations (timeout can be set only once per app!).

0xc0170 committed:
Watchdog: start without arg uses max timeout from a target
